src/group_theory/sylow.lean

```diff
@@ -17,21 +17,146 @@ The Sylow theorems are the following results for every finite group `G` and ever
   `p`-subgroup, `nₚ ≡ 1 [MOD p]`, and `nₚ` is equal to the index of the normalizer of the Sylow
   `p`-subgroup in `G`.
 
-In this file, currently only the first of these results is proven.
+## Main definitions
+
+* `sylow p G` : The type of Sylow `p`-subgroups of `G`.
 
 ## Main statements
 
-* `exists_subgroup_card_pow_prime`: A generalisation of the first of the Sylow theorems: For every
-  prime power `pⁿ` dividing `G`, there exists a subgroup of `G` of order `pⁿ`.
+* `exists_subgroup_card_pow_prime`: A generalization of Sylow's first theorem:
+  For every prime power `pⁿ` dividing the cardinality of `G`,
+  there exists a subgroup of `G` of order `pⁿ`.
+* `is_p_group.exists_le_sylow`: A generalization of Sylow's first theorem:
+  Every `p`-subgroup is contained in a Sylow `p`-subgroup.
+* `sylow_conjugate`: A generalization of Sylow's second theorem:
+  If the number of Sylow `p`-subgroups is finite, then all Sylow `p`-subgroups are conjugate.
+* `card_sylow_modeq_one`: A generalization of Sylow's third theorem:
+  If the number of Sylow `p`-subgroups is finite, then it is congruent to `1` modulo `p`.
+-/
+
+open fintype mul_action subgroup
 
-## TODO
+section infinite_sylow
 
-* Prove the second and third of the Sylow theorems.
-* Sylow theorems for infinite groups
--/
+variables (p : ℕ) (G : Type*) [group G]
+
+/-- A Sylow `p`-subgroup is a maximal `p`-subgroup. -/
+structure sylow extends subgroup G :=
+(is_p_group' : is_p_group p to_subgroup)
+(is_maximal' : ∀ {Q : subgroup G}, is_p_group p Q → to_subgroup ≤ Q → Q = to_subgroup)
+
+variables {p} {G}
+
+instance : has_coe (sylow p G) (subgroup G) := ⟨sylow.to_subgroup⟩
+
+@[ext] lemma sylow.ext {P Q : sylow p G} (h : (P : subgroup G) = Q) : P = Q :=
+by cases P; cases Q; congr'
+
+lemma sylow.ext_iff {P Q : sylow p G} : P = Q ↔ (P : subgroup G) = Q :=
+⟨congr_arg coe, sylow.ext⟩
+
+/-- A generalization of **Sylow's first theorem**.
+  Every `p`-subgroup is contained in a Sylow `p`-subgroup. -/
+lemma is_p_group.exists_le_sylow {P : subgroup G} (hP : is_p_group p P) :
+  ∃ Q : sylow p G, P ≤ Q :=
+exists.elim (zorn.zorn_nonempty_partial_order₀ {Q : subgroup G | is_p_group p Q} (λ c hc1 hc2 Q hQ,
+⟨ { carrier := ⋃ (R : c), R,
```

Comment:
Might be easier as
```suggestion
⟨ { carrier := Sup c,
```
Unless there's a coercion I'm missing

Reply:
There's a hidden coercion from subgroup to sets.